### PR TITLE
RPC support in the example tool chef linux build

### PR DESCRIPTION
**Pull request by MtTsai**
Port RPC server enabled logic into example tool chef, and also all the
pre-configured zap file for different device types

the usage example:
  $ ./chef.py -zbr -t linux -d lighting --pid 0x8000

#### Problem
What is being fixed?  Examples:
* No RPC server support in the linux build of example tool chef

#### Change overview
* Add RPC server support in the linux build of example tool chef
* Add pre-configured zap files for different device types in the linux build of example tool chef

#### Testing
How was this tested? (at least one bullet point required)
* If unit tests were added, how do they cover this issue?
* If unit tests existed, how were they fixed/modified to prevent this in future?
* If new unit tests are not added, why not?
* If integration tests were added, how do they verify this change?
* If new integration tests are not added, why not?
* If manually tested, what platforms controller and device platforms were manually tested, and how? 
  * follow the commands to build and run the virtual device on linux with RPC support
```
$ cd examples/chef
$ ./chef.py --bootstrap_zap
$ ./chef.py -zbr -t linux -d lighting-app --pid 0x8000
$ ./linux/out/lighting
```
* If no testing is required, why not?

### CODE REVIEW

File: examples/chef/chef_config.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+nrfconnectFolder = "/opt/nordic/ncs/"
+#nrfConnectSerialDevice = "/dev/tty.usbmodem0*" # Typical MacOS path
+nrfConnectSerialDevice = "/dev/ttyACM0"  # Typical Linux path
+
+# ESP32
+
+esp32Folder = os.path.expandvars("$HOME/tools/esp-idf")
+esp32SerialDevice = "/dev/ttyUSB0"  # Typical Linux path
+#esp32SerialDevice = "/dev/tty.usbserial-*" # Typical MacOS path
```
[the remark this answers is not in the document]
Comment (rochaferraz) [reply examples/chef/chef_config.py:18]: +1. This is not used anymore

File: examples/chef/devices/bridge-app.zap [hunk 2]
```diff
@@ -0,0 +1,9473 @@
+{
```
Comment (rochaferraz) [inline examples/chef/devices/bridge-app.zap:1]: Have you tested these device configurations at master?

File: examples/chef/chef_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MtTsai) [reply examples/chef/chef_config.py:18]: OK, I'll delete it, thanks!

File: examples/chef/linux/with_pw_rpc.gni [hunk 3]
```diff
@@ -0,0 +1 @@
+../../lighting-app/linux/with_pw_rpc.gni
```
[the remark this answers is not in the document]
Comment (MtTsai) [reply examples/chef/linux/with_pw_rpc.gni:1]: OK, I'll make a copy, thanks!

File: examples/chef/devices/bridge-app.zap (hunk 2 above)
Comment (MtTsai) [reply examples/chef/devices/bridge-app.zap:1]: I've only tested for lighting-app, I'll keep only this one in the change.

File: examples/chef/linux/BUILD.gn [hunk 4]
```diff
@@ -12,38 +12,91 @@
...
-  zap_pregenerated_dir =
-      "${chip_root}/examples/chef/out/${sample_name}/zap-generated/"
+  zap_pregenerated_dir = "${chip_root}/examples/chef/zap-generated"
   is_server = true
 }
 
 executable("${sample_name}") {
-  sources = [ "${project_dir}/linux/main.cpp" ]
+  sources = [
+    "${project_dir}/common/stubs.cpp"
```
[the remark this answers is not in the document]
Comment (MtTsai) [reply examples/chef/linux/BUILD.gn]: Thanks for catching this.

File: examples/chef/linux/BUILD.gn [hunk 5]
```diff
@@ -12,38 +12,91 @@
...
+import("sample.gni")
+
 project_dir = "./.."
 
 chip_data_model("chef-data-model") {
   zap_file = "${project_dir}/devices/${sample_zap_file}"
-
-  zap_pregenerated_dir =
-      "${chip_root}/examples/chef/out/${sample_name}/zap-generated/"
+  zap_pregenerated_dir = "${chip_root}/examples/chef/zap-generated"
```
[the remark this answers is not in the document]
Comment (MtTsai) [reply examples/chef/linux/BUILD.gn]: yes, you are right!

**Comment by bzbarsky-apple** [comment]
> Why does this work for Linux and not Mac?

It might depend on the compiler version, and on gcc vs clang....

Does disabling this warning in `disabled_warnings_third_party` in [build/config/compiler/BUILD.gn](https://github.com/project-chip/connectedhomeip/blob/master/build/config/compiler/BUILD.gn#L242) help?

If not, does disabling it in the cflags for the executable in examples/chef/linux/BUILD.gn work?

**Comment by MtTsai** [comment]
> 

The flags in `disabled_warnings_third_party` couldn't be applied for building pw_string/format.cc.
And I'm not sure if there is a way to make it.

Add cflags for the executable in examples/chef/linux/BUILD.gn doesn't be applied for building pw_string/format.cc, too.

**Comment by MtTsai** [comment]
I uploaded this change with new pigweed reference, which fixed the warning for format on Darwin system.
Therefore, this cl could be built pass on Darwin successfully now.

**Review by cpagravel — APPROVED** [review]

File: examples/chef/common/stubs.cpp [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#include <app-common/zap-generated/attribute-id.h>
+#include <app-common/zap-generated/attribute-type.h>
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/callback.h>
+#include <app-common/zap-generated/cluster-id.h>
+#include <app-common/zap-generated/command-id.h>
+
+bool emberAfPluginDoorLockOnDoorLockCommand(chip::EndpointId endpointId, chip::Optional<chip::ByteSpan> pinCode)
```
Comment (andy31415) [inline examples/chef/common/stubs.cpp:8]: why is door lock special here? why is the file called stubs.cpp not door_lock_stubs.cpp?

Some comment explaining why this file exist and what it is supposed to contain helps here. if this is named 'door_lock_stubs.cpp' that would not be needed, but if stubs it does make me wonder why lock is here but not others.

File: examples/chef/chef.py [hunk 7]
```diff
@@ -333,8 +333,12 @@ def main(argv: Sequence[str]) -> None:
             with open(f"{_CHEF_SCRIPT_PATH}/linux/args.gni", "w") as f:
                 f.write(textwrap.dedent(f"""\
                         import("//build_overrides/chip.gni")
-                        import("\\${{chip_root}}/config/standalone/args.gni")
+                        import("${{chip_root}}/config/standalone/args.gni")
                         chip_shell_cmd_server = false
+                        chip_build_libshell = true
+                        chip_config_network_layer_ble = false
+                        chip_enable_wifi=false
+                        chip_mdns="minimal"
```
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/chef/chef.py:341]: If this is "make mac and linux behave similarely" that is fair ... but also wondering why we do this. If mac has a specific mdns that it wants to use by default, why does chef want to behave differently.

[the remark this answers is not in the document]
Comment (MtTsai) [reply examples/chef/chef.py:341]: Thanks for catching these, I think they are out-dated overriden arguments.
It's not necessary to do it in this change. I'll remove them.

**Review by andy31415 — APPROVED** [review]

**Comment by MtTsai** [comment]
> @MtTsai - pulling the latest pigweed does not work - see #18785

As discussion, we could put the change with no pigweed update for now.

**Pull request merged**